### PR TITLE
refine multi uprobe handling

### PR DESCRIPTION
**Pull request by swananan**
Follow up to #1417.
I responded to the latest review comments in #1417.

### Added/updated tests?

_We strongly encourage you to add a test for your changes._

- [x] No, and this is why: the existed tests have covered the feature.

### Checklist

- [x] Rust code has been formatted with `cargo +nightly fmt`.
- [x] All clippy lints have been fixed.
      You can find failing lints with `cargo xtask clippy`.
- [x] Unit tests are passing locally with `cargo test`.
- [x] The [Integration tests] are passing locally.
- [x] I have blessed any API changes with `cargo xtask public-api --bless`.

[Integration tests]: https://github.com/aya-rs/aya/blob/main/test/README.md

### (Optional) What GIF best describes this PR or how it makes you feel?

- - -
This change is [Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)

### CODE REVIEW

**Review by tamird — COMMENTED** [review]
> @tamird reviewed 19 files and all commit messages, and made 3 comments.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: all files reviewed, 3 unresolved discussions (waiting on swananan).

___
*[`aya/src/programs/probe.rs` line 112 at r1](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz3czLdDp7IuLwVnuUk:-Oz3czLdDp7IuLwVnuUl:b-wh9of0) ([raw file](https://github.com/aya-rs/aya/blob/45ea5837a9e7753f73ce8298ecf9054146669337/aya/src/programs/probe.rs#L112)):*
> ```XML
>             Self::Many(links) => {
>                 for link in links {
>                     link.detach()?;
> ```

I think you can write something like `let Ok(()) = link.detach();` to ensure this doesn't regress

probably above as well 
___
*[`aya/src/programs/uprobe.rs` line 540 at r1](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz3dG3kBXLqNeGHyuEb:-Oz3dG3kBXLqNeGHyuEc:b-2guby3) ([raw file](https://github.com/aya-rs/aya/blob/45ea5837a9e7753f73ce8298ecf9054146669337/aya/src/programs/uprobe.rs#L540)):*
> ```XML
>                             // Roll back all successfully attached points before returning the
>                             // original attach error.
>                             let _unused: Result<(), Infallible> =
> ```

same here
___
*[`aya/src/programs/uprobe.rs` line 88 at r3](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz3d_xe0-sy2QEss_Yj:-Oz3d_xe0-sy2QEss_Yk:bm0xu78) ([raw file](https://github.com/aya-rs/aya/blob/b0e208f96cdcd914b87d0b6709c83953d264b938/aya/src/programs/uprobe.rs#L88)):*
> ```XML
> }
> 
> // Allows borrowed symbol collections, such as `&[&str]`, to be passed directly.
> ```

i don't get it, why not do a blanket impl for any reference to something that impls the trait?

**Review by tamird — COMMENTED** [review]
> @tamird made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: all files reviewed, 4 unresolved discussions (waiting on swananan).

___
*[`aya/src/programs/uprobe.rs` line 278 at r3](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz3fGzpCNoR7gr3wyNS:-Oz3fGzpCNoR7gr3wyNT:b4k8y8p) ([raw file](https://github.com/aya-rs/aya/blob/b0e208f96cdcd914b87d0b6709c83953d264b938/aya/src/programs/uprobe.rs#L278)):*
> ```XML
> // target and optional debug object are each mapped and scanned at most once per attach call.
> // `PendingOffset` preserves mixed absolute and symbolic locations plus symbol addends, allowing
> // offsets to be rebuilt in caller order and remain aligned with cookies.
> ```

maybe implementation comments would be better inside.



_Code quote:_
```XML
// Resolve points without losing their caller-defined order.
//
// A single point takes a direct path to avoid the Vec and HashMap used for batch resolution.
// Multiple points are collected first, then their unique symbols are resolved together so the
// target and optional debug object are each mapped and scanned at most once per attach call.
// `PendingOffset` preserves mixed absolute and symbolic locations plus symbol addends, allowing
// offsets to be rebuilt in caller order and remain aligned with cookies.
```

**Review by swananan — COMMENTED** [review]
> @swananan made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: 12 of 19 files reviewed, 4 unresolved discussions (waiting on tamird).

___
*[`aya/src/programs/probe.rs` line 112 at r1](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz3czLdDp7IuLwVnuUk:-Oz4SpHY56Pv0cKZfo5E:b4grwqd) ([raw file](https://github.com/aya-rs/aya/blob/45ea5837a9e7753f73ce8298ecf9054146669337/aya/src/programs/probe.rs#L112)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think you can write something like `let Ok(()) = link.detach();` to ensure this doesn't regress

probably above as well
</blockquote></details>

Done. I checked the Infallible documentation (https://doc.rust-lang.org/std/convert/enum.Infallible.html), but it doesn’t mention this pattern. I am planning to open a PR to enrich doc, which looks like a good first contribution for me. 
___
*[`aya/src/programs/uprobe.rs` line 540 at r1](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz3dG3kBXLqNeGHyuEb:-Oz4U0gzAx-tj9URQ74M:b-896fix) ([raw file](https://github.com/aya-rs/aya/blob/45ea5837a9e7753f73ce8298ecf9054146669337/aya/src/programs/uprobe.rs#L540)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Done.
___
*[`aya/src/programs/uprobe.rs` line 88 at r3](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz3d_xe0-sy2QEss_Yj:-Oz4_25-BOiguoqdrhls:b-896fix) ([raw file](https://github.com/aya-rs/aya/blob/b0e208f96cdcd914b87d0b6709c83953d264b938/aya/src/programs/uprobe.rs#L88)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i don't get it, why not do a blanket impl for any reference to something that impls the trait?
</blockquote></details>

Done.
___
*[`aya/src/programs/uprobe.rs` line 278 at r3](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz3fGzpCNoR7gr3wyNS:-Oz4X478DGLZFvMEds9H:b-896fix) ([raw file](https://github.com/aya-rs/aya/blob/b0e208f96cdcd914b87d0b6709c83953d264b938/aya/src/programs/uprobe.rs#L278)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

maybe implementation comments would be better inside.
</blockquote></details>

Done.

**Review by tamird — COMMENTED** [review]
> @tamird reviewed 7 files and all commit messages, made 5 comments, and resolved 4 discussions.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: all files reviewed, 4 unresolved discussions (waiting on swananan).

___
*[`aya/src/programs/probe.rs` line 112 at r1](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz3czLdDp7IuLwVnuUk:-Oz723ZOD7iEODUHySWG:bb7lbs2) ([raw file](https://github.com/aya-rs/aya/blob/45ea5837a9e7753f73ce8298ecf9054146669337/aya/src/programs/probe.rs#L112)):*
<details><summary><i>Previously, swananan (swananan) wrote…</i></summary><blockquote>

Done. I checked the Infallible documentation (https://doc.rust-lang.org/std/convert/enum.Infallible.html), but it doesn’t mention this pattern. I am planning to open a PR to enrich doc, which looks like a good first contribution for me.
</blockquote></details>

nice
___
*[`aya/src/programs/probe.rs` line 91 at r7](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz73gkTE_NC-g80YCnp:-Oz73gkTE_NC-g80YCnq:b12h5kj) ([raw file](https://github.com/aya-rs/aya/blob/fd6ccad331f2a0011a91ca1f3c5e16e1a2e5da6d/aya/src/programs/probe.rs#L91)):*
> ```XML
>                     return Err(ProgramError::AttachCookieNotSupported);
>                 }
>                 // Both paths attach with PERF_EVENT_IOC_SET_BPF. The kernel version only
> ```

feels like we're duplicating something here
___
*[`aya/src/programs/uprobe.rs` line 93 at r6](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz72FFkADzCjGmRhpyO:-Oz72FFkADzCjGmRhpyP:b-t82ybg) ([raw file](https://github.com/aya-rs/aya/blob/eea45b7288720ca6771d2e75106755c0c235f382/aya/src/programs/uprobe.rs#L93)):*
> ```XML
> }
> 
> // Allow iterators over borrowed collections to reuse conversions implemented for their elements.
> ```

i think you can just drop this comment? no longer about iterators
___
*[`aya/src/programs/uprobe.rs` line 538 at r7](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz72yh48AcHVHJpn1DZ:-Oz72yh48AcHVHJpn1D_:b-buyixd) ([raw file](https://github.com/aya-rs/aya/blob/fd6ccad331f2a0011a91ca1f3c5e16e1a2e5da6d/aya/src/programs/uprobe.rs#L538)):*
> ```XML
>                 let prog_fd = data.fd()?;
>                 let prog_fd = prog_fd.as_fd();
>                 let Some(&first_offset) = offsets.first() else {
> ```

something here needs commentary. saving allocations?
___
*[`aya/src/programs/uprobe.rs` line 569 at r7](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz73FfM54vFm3W9RFbO:-Oz73FfM54vFm3W9RFbP:b-g6ow3o) ([raw file](https://github.com/aya-rs/aya/blob/fd6ccad331f2a0011a91ca1f3c5e16e1a2e5da6d/aya/src/programs/uprobe.rs#L569)):*
> ```XML
>                         .as_ref()
>                         .and_then(|cookies| cookies.get(index).copied());
>                     if let Err(error) =
> ```

if let err makes it hard to see what the ok variant is and why it's ok to drop it

**Review by swananan — COMMENTED** [review]
> @swananan made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: all files reviewed, 4 unresolved discussions (waiting on tamird).

___
*[`aya/src/programs/probe.rs` line 91 at r7](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz73gkTE_NC-g80YCnp:-Oz7EBMF9h4hWKaCVTUr:b-b5qfn7) ([raw file](https://github.com/aya-rs/aya/blob/fd6ccad331f2a0011a91ca1f3c5e16e1a2e5da6d/aya/src/programs/probe.rs#L91)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

feels like we're duplicating something here
</blockquote></details>

Done. I was concerned that the meaning of kernel version check was not clear, but this comment is not very helpful, so I removed it.
___
*[`aya/src/programs/uprobe.rs` line 93 at r6](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz72FFkADzCjGmRhpyO:-Oz7BbKr2BFKjf2Qve2c:b-896fix) ([raw file](https://github.com/aya-rs/aya/blob/eea45b7288720ca6771d2e75106755c0c235f382/aya/src/programs/uprobe.rs#L93)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think you can just drop this comment? no longer about iterators
</blockquote></details>

Done.
___
*[`aya/src/programs/uprobe.rs` line 538 at r7](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz72yh48AcHVHJpn1DZ:-Oz7HZfl9oG3Ue5oYpsJ:blho77c) ([raw file](https://github.com/aya-rs/aya/blob/fd6ccad331f2a0011a91ca1f3c5e16e1a2e5da6d/aya/src/programs/uprobe.rs#L538)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

something here needs commentary. saving allocations?
</blockquote></details>

It's not about saving allocations, I added a comment to explain why.
___
*[`aya/src/programs/uprobe.rs` line 569 at r7](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz73FfM54vFm3W9RFbO:-Oz7I8in4ekWSoVcEPEu:b-896fix) ([raw file](https://github.com/aya-rs/aya/blob/fd6ccad331f2a0011a91ca1f3c5e16e1a2e5da6d/aya/src/programs/uprobe.rs#L569)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

if let err makes it hard to see what the ok variant is and why it's ok to drop it
</blockquote></details>

Done.

**Review by tamird — COMMENTED** [review]
> @tamird reviewed 7 files and all commit messages, made 1 comment, and resolved 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: all files reviewed, 1 unresolved discussion (waiting on swananan).

___
*[`aya/src/programs/probe.rs` line 91 at r7](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz73gkTE_NC-g80YCnp:-Oz7OSM03x6iwBMqi0oR:b-c8467l) ([raw file](https://github.com/aya-rs/aya/blob/fd6ccad331f2a0011a91ca1f3c5e16e1a2e5da6d/aya/src/programs/probe.rs#L91)):*
<details><summary><i>Previously, swananan (swananan) wrote…</i></summary><blockquote>

Done. I was concerned that the meaning of kernel version check was not clear, but this comment is not very helpful, so I removed it.
</blockquote></details>

but isn't this code duplicated somewhere?

**Review by swananan — COMMENTED** [review]
> @swananan made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: 15 of 19 files reviewed, 1 unresolved discussion (waiting on tamird).

___
*[`aya/src/programs/probe.rs` line 91 at r7](https://reviewable.io/reviews/aya-rs/aya/1654#-Oz73gkTE_NC-g80YCnp:-Oz9k0D32jkfm6pE_c2C:byqzwny) ([raw file](https://github.com/aya-rs/aya/blob/fd6ccad331f2a0011a91ca1f3c5e16e1a2e5da6d/aya/src/programs/probe.rs#L91)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

but isn't this code duplicated somewhere?
</blockquote></details>

Oh, got it, I made a small refactor. The kernel version check is defined in the probe_pmu_supported(). There are 3 distinct attachment in aya:

1. TracePoint and PerfEvent use perf_attach(), but they don't need to do the kernel check.
2. Normal uprobe and kprobe still use the attach_impl, but they doesn't use perf_attach() any more. On the perf-event fallback path, probe_pmu_supported will execute twice. The check is not expensive, so I think that's acceptable.
3. For the per-point Many path, the subsequent attachments call the specific helper instead of using perf_attach().

**Review by tamird — APPROVED** [review]
> @tamird reviewed 4 files and all commit messages, made 1 comment, and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: all files reviewed, 1 unresolved discussion (waiting on swananan).

___
*[`aya/src/programs/probe.rs` line 84 at r12](https://reviewable.io/reviews/aya-rs/aya/1654#-OzC5dPUDtuA6b1iHxJ7:-OzC5dPUDtuA6b1iHxJ8:b-unupw3) ([raw file](https://github.com/aya-rs/aya/blob/c89c70d4199f2f1c60e4670b3fc3e369e128758e/aya/src/programs/probe.rs#L84)):*
> ```XML
>         // fd-backed or perf-backed path directly without repeating backend selection.
>         match self {
>             Self::Fd(links) => links.push(attach_bpf_probe::<P>(
> ```

move the `links.push` call out of the match, that will also make the error handling a bit clearer (`?` can be on the outside `let ... = match { ... }?;`

**Review by swananan — COMMENTED** [review]
> @swananan made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: 15 of 19 files reviewed, 1 unresolved discussion (waiting on tamird).

___
*[`aya/src/programs/probe.rs` line 84 at r12](https://reviewable.io/reviews/aya-rs/aya/1654#-OzC5dPUDtuA6b1iHxJ7:-OzCR_yh1M5wnzOGxyzd:b824n24) ([raw file](https://github.com/aya-rs/aya/blob/c89c70d4199f2f1c60e4670b3fc3e369e128758e/aya/src/programs/probe.rs#L84)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

move the `links.push` call out of the match, that will also make the error handling a bit clearer (`?` can be on the outside `let ... = match { ... }?;`
</blockquote></details>

Links.push can not be moved out of the match, since it has different link types. Instead, I created link first, which should be more readable.

**Review by tamird — APPROVED** [review]
> @tamird reviewed 4 files and all commit messages, and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1654)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on swananan).

**Comment by tamird** [comment]
@codex review
